src/modules/roc_netio/target_uv/roc_netio/transceiver.cpp

```diff
@@ -214,76 +205,82 @@ void Transceiver::stop_sem_cb_(uv_async_t* handle) {
     roc_panic_if_not(handle);
 
     Transceiver& self = *(Transceiver*)handle->data;
-    self.stop_();
-    self.close_();
+    self.async_close_ports_();
+    self.close_sems_();
     self.process_tasks_();
 }
 
+void Transceiver::close_sems_cb_(uv_handle_t* handle) {
+    roc_panic_if_not(handle);
+
+    Transceiver& self = *(Transceiver*)handle->data;
+
+    core::Mutex::Lock lock(self.mutex_);
+
+    if (handle == (uv_handle_t*)&self.stop_sem_) {
+        self.stop_sem_initialized_ = false;
+    } else {
+        self.task_sem_initialized_ = false;
+    }
+
+    if (self.stop_sem_initialized_ || self.task_sem_initialized_) {
+        return;
+    }
+
+    self.cond_.broadcast();
+}
+
 void Transceiver::stop_() {
-    for (core::SharedPtr<UDPReceiver> rp = receivers_.front(); rp;
-         rp = receivers_.nextof(*rp)) {
-        rp->stop();
+    core::Mutex::Lock lock(mutex_);
+
+    if (stopped_) {
+        return;
     }
 
-    for (core::SharedPtr<UDPSender> sp = senders_.front(); sp;
-         sp = senders_.nextof(*sp)) {
-        sp->stop();
+    if (int err = uv_async_send(&stop_sem_)) {
+        roc_panic("transceiver: uv_async_send(): [%s] %s", uv_err_name(err),
+                  uv_strerror(err));
     }
+
+    wait_stopped_();
+    stopped_ = true;
```

Comment:
It seems that there is a problem if stop() will be called to times concurrently:

* thread 1 acquires the mutex
* thread 2 sleeps on the mutex
* thread 1 calls uv_async_send() and calls wait_stopped_(), which temporary releases the mutex and goes to sleep
* thread 2 wakes up, acquires the mutex, and calls uv_async_send()

This way, uv_async_send() is called twice. But stop_sem_cb_() is not intended to be called more than once. For example, close_sems_() will incorrectly call uv_close() more than once.

Reply:
BTW, why at all we need the wait_stopped_() call? We anyway call Thread::join() after calling stop_() so it looks unnecessary. If we remove wait_stopped_() the problem will gone.

Reply:
>  why at all we need the wait_stopped_() call

It is an ancestor of the time when `trx.stop()` did not wait the background thread to be finished.

Reply:
Reworked